### ISSUE TITLE
XML type manifold support

### ISSUE THREAD

**Issue by rsmckinney** [manifold-systems/manifold#130]
This issue continues the discussion from https://github.com/manifold-systems/manifold/issues/111, focusing more on XML type manifolds rather than the XSD manifold.
 
**Update**
What has developed from this discussion is an "XML Manifold", which is essentially XML schema *by-example*, which is similar to the [JSON Manifold](https://github.com/manifold-systems/manifold/tree/master/manifold-deps-parent/manifold-json) where it supports both by-example JSON and JSON Schema. In fact, as with the [YAML Manifold](https://github.com/manifold-systems/manifold/tree/master/manifold-deps-parent/manifold-yaml), the XML Manifold amounts to a thin layer over the core JSON framework.  Essentially it transforms XML to/from JSON bindings the same way YAML does.  This way you can use JSON, YAML, and XML interchangeably and equally leverage all the features from the JSON framework -- knowing one means you know them all.

**Comment by sporkisfaster** [comment]
So I assume the approach is the following: https://github.com/manifold-systems/manifold/tree/master/manifold-core-parent/manifold#anatomy-of-a-type-manifold are there any special xml considerations? Tips?

**Comment by rsmckinney** [comment]
Generally, yes.  Although, given this is a somewhat common use-case, I think it would be best if I were to build a small foundation for this use case, probably a base class and supporting model. But before I do this, it'd be nice to have your use-case fleshed out a little bit: what do the xml files look like? how are they named? how are they being used? etc.  My idea of "common" may not fit your use-case.

**Comment by sporkisfaster** [comment]
Of course, let me scrub one of them and I'll post it here.

**Comment by sporkisfaster** [comment]
Given the below XML snippet (which is in a file). We are trying to more cleanly grab particular nodes and use them as a basis for a test using selenium. So we do things like "if action is ENTER do a thing, then check any screens/fields after this action if they are valid". Or sometimes we don't want fields that have a particular attribute. Or skip screens that have a certain ID and so on.
`<flow>
    <events>
        <screen type="RECEIVE" id="0">
            <field readwrite="read" visible="true" row="38" col="0">
                <text>TITLES RULE</text>
            </field>
            <field readwrite="read" visible="true" row="39" col="1">
                <text>SO DO PROMPTS</text>
            </field>
            <field readwrite="read" visible="true" row="40" col="1">
                <text>AND OTHER WEB DOODLES</text>
            </field>
            <field readwrite="read" visible="true" row="41" col="0">
                <text>ETC</text>
            </field>
            <field readwrite="read" visible="true" row="42" col="0">
                <text>MORE TEXT</text>
            </field>
            <field readwrite="read" visible="true" row="43" col="0">
                <text>ALMOST DONE</text>
            </field>
            <field readwrite="read" visible="true" row="44" col="0">
                <text>SO CLOSE</text>
            </field>
            <field readwrite="read" visible="true" row="45" col="0">
                <text>YEET</text>
            </field>
            <field readwrite="read" visible="true" row="46" col="0">
                <text></text>
            </field>
        </screen>
        <action id="1">
            <key>ENTER</key>
        </action>
    </events>
</flow>`

**Comment by rsmckinney** [comment]
Does this data correspond with another type or types you are testing? For instance, is there a "screen" object or a type for a given screen?  In other words, would it be more useful if this data could augment and existing type, or is it better mapped to a type of its own?

**Comment by rsmckinney** [comment]
Seeing how you are currently using this would help, but I understand if you can't disclose that.

**Comment by sporkisfaster** [comment]
As of right now no it's not feeding into any objects of our purpose. Mostly data to go into selenium driver. So no marshaling/marshaling just pure xml data.

**Comment by sporkisfaster** [comment]
We can certainly achieve this with various parsers but they get a little unreadable when we start trying to get more specific about which child nodes we want or want to modify.

**Comment by sporkisfaster** [comment]
Also the files themselves get huge and are numerous, so running logic and loading them ALL fully into memory sucks sometimes.

**Comment by rsmckinney** [comment]
Ok, I think I have a pretty good idea of your process.  This matches my general mental model, which helps to validate the foundation I'll start building now.

**Comment by rsmckinney** [comment]
Ok, one more question. The XML files, where are they on disk? How are they named? Are they available when you compile your tests?

**Comment by sporkisfaster** [comment]
So they will generally either be in a resources folder under main or test. OR have the option to specifically point to a directory or subdirectory they choose. Usually the former, sometimes latter.

**Comment by rsmckinney** [comment]
How would you expect to use a file?  For example, given file `resources/com/example/testfiles/Alpha.xml` with the format from above, a generalized solution would look like:
```java
import com.example.testfiles.Alpha;

if (Alpha.events.screen.action.key.equals("ENTER")) {
  // do a thing
  if (Alpha.events.screen.field.get(0).isReadwrite()) ...
}
```
Notice there is a type corresponding with the `Alpha.xml` file. This can work, however, in light of many more files having the exact same structure it may be more beneficial to instead define a "by-example" file that defines the type; you could use one of your existing files for the example provided it is a *complete* sample.  This way you can have a single type, say `OmegaTest`, where you'd use it like this:
```java
import com.example.types.OmegaTest;

OmegaTest alpha = OmegaTest.from("/path/to/test/Alpha.xml");

if (alpha.events.screen.action.key.equals("ENTER")) {
  // do a thing
  if (alpha.events.screen.field.get(0).isReadwrite()) ...
}
```
One significant benefit from this approach is you can leverage the crap out of **extension methods**:
```java
@Extension
public class MyOmegaTestExtension {
  public static boolean isEnter(@This OmegaTest thiz) {
    return thiz.events.screen.action.key.equals("ENTER");
  }
. . . // all kinds of other useful methods
}
```
Now you can write code like this:
```jav
if (alpha.isEnter()) {
  // do a thing
}
```

**Comment by sporkisfaster** [comment]
So this looks excellent. The only things are: each xml may have a different number of elements (screens, actions, fields) but their composition will always be the same (attributes in those elements).  Not sure if and how this changes anything above. Lastly what about about attribute access and manipulation? I think I see that with is read write but unsure. Could I add a whole new attribute like add(canSkip = true)

**Comment by sporkisfaster** [comment]
Also say what if I have a directory full of xmls with different file names. How would that work in the above example

**Comment by rsmckinney** [comment]
The `OmegaTest` type loads instances from the files:
```java
OmegaTest test1 = OmegaTest.from("/path/to/test/Test1.xml");
OmegaTest test2 = OmegaTest.from("/path/to/test/Test2.xml");
OmegaTest test3 = OmegaTest.from("/path/to/test/Test3.xml");
```
Again, you can use extension methods to make this easier for your specific environment:
```java
@Extension
public class MyOmegaTestExtension {
  @Extension
  public static OmegaTest from(String name) {
    return OmegaText.from(findTestPath(name));
  }
  
  private static String findTestPath(String name) {
    . . . 
  }
}
```
So you can write:
```java
OmegaTest test1 = OmegaTest.from("Test1.xml");
OmegaTest test2 = OmegaTest.from("Test2.xml");
OmegaTest test3 = OmegaTest.from("Test3.xml");
```
This assumes the files are uniformly structured and can be represented with the `OmegaTest.xml` sample, which appears to be the case.

**Comment by rsmckinney** [comment]
Re
>Lastly what about about attribute access and manipulation? I think I see that with is read write but unsure. Could I add a whole new attribute like add(canSkip = true)

Yes, elements and attributes are read/write, which will be configurable.  Also configurable is dynamic addition/removal of structure e.g., `test1.events.screen.put("canSkip", true)`

**Comment by sporkisfaster** [comment]
Fantastic

**Comment by rsmckinney** [comment]
Experimental release of [XML Manifold](https://github.com/manifold-systems/manifold/tree/master/manifold-deps-parent/manifold-xml) is now available with release `2019.1.23`.

Wes, you can use the `manifold-xml` dependency mostly as discussed here, but with improvements that are consistent with the existing JSON and YAML manifolds.  For instance, you can treat any of your files as a type, if that's what you prefer:
```java
// types-safely use the XML resource file: resources/com/example/testfiles/MyFile.xml
import com.example.testfiles.MyFile;

// load the data in the file into an instance of MyFile
MyFile myFile = MyFile.fromSource();

// Use it type-safely
var screen = myFile.getEvents().getScreen();
etc.
```
Or you can derive a sample file representing a the common structure of all the test files:
```java
// types-safely use a "universal" sample test type
import com.example.testfiles.OmegaTest;

// create an instance of OmegaTest from any a test file
OmegaTest myFile = OmegaTest.load().fromXmlUrl(testResourceUrl);

// Use it type-safely
var screen = myFile.getEvents().getScreen();
etc.
```
Using the latter technique you can create extension methods on `OmegaTest` to improve usability matching your test use-cases, if that helps.

**Comment by sporkisfaster** [comment]
Can’t wait to try it out. Thanks so much for your quick and hard work on this

**Comment by rsmckinney** [comment]
No problem. XML support was long overdue.

Just finished releasing Manifold `2019.1.25` and a new IJ plugin release with fixes to fully support XML:
- Navigate from/to Java ode usages of XML elements/attributes
- Find usages of XML element/attributes in Java code
- Rename/refactor of XML elements/attributes (changes Java code refs)

Note XML, JSON, and YAML are interchangeable.  This means there is one universal API, which is documents in the [JSON Manifold](https://github.com/manifold-systems/manifold/tree/master/manifold-deps-parent/manifold-json).  Let me know if you have questions.

Cheers!

**Comment by rsmckinney** [comment]
Released `2019.1.26`.  Fixes:
- https://github.com/manifold-systems/manifold/issues/134
- https://github.com/manifold-systems/manifold/issues/135
- some other minor fixes
